src/libexpr-c/nix_api_value.cc

```diff
@@ -112,6 +126,7 @@ ValueType nix_get_type(nix_c_context * context, const Value * value)
         context->last_err_code = NIX_OK;
     try {
         auto & v = check_value_not_null(value);
+        check_value_initialized(v);
```

Comment:
`check_value_initialized` is only used after `check_value_not_null`, which suggests that we should perhaps have a helper function that does both?
We can then also rephrase it in higher level terms:
`check_value_in`: `check_value_not_null` + `check_value_initialized`
`check_value_out`: `check_value_not_null` + `check_value_uninitialized`

The only other cases I can think of now, where we might want to use `check_value_initialized` separately is
- nullable parameters. For those we could have `check_value_in_or_null`.
- checking multiple values in a buffer, e.g. list or attrset builder. I'm not entirely sure we should do that, because it makes constructing cyclic structures difficult.

Reply:
Makes sense, I added `check_value_in` and `check_value_out` helpers. I'm not happy about having separate versions for `const` and `non-const` data, but I can't think of a good solution.

---

> nullable parameters. For those we could have `check_value_in_or_null`.

Good point. However, I don't believe we currently have any functions that accept nullable values. If necessary, we can add a helper function for that in the future.
---



> checking multiple values in a buffer, e.g. list or attrset builder. I'm not entirely sure we should do that, because it makes constructing cyclic structures difficult.

In addition, the order of calling `nix_init_*` and `nix_[list,builder]_builder_insert` shouldn't matter. I've updated these functions to only check for a null value.

Reply:
@roberth Kindly pinging you, since I think you might have missed the latest changes. (If not, no worries, just wanted to make sure you're aware, and take your time to review it)